system/jlib/jevent.cpp

```diff
@@ -0,0 +1,560 @@
+    offset_type writeOffset = reserveEvent(requiredSize);
+    offset_type pos = writeOffset;
+    writeEventHeader(EventIndexLoad, pos);
+    write(pos, EvAttrFileId, fileid);
+    write(pos, EvAttrFileOffset, offset);
+    write(pos, EvAttrNodeKind, nodeKind);
+    write(pos, EvAttrExpandedSize, size);
+    write(pos, EvAttrElapsedTime, elapsedTime);
+    write(pos, EvAttrReadTime, readTime);
+    writeEventFooter(pos, requiredSize, writeOffset);
+}
+
+void EventRecorder::recordIndexEviction(unsigned fileid, offset_t offset, byte nodeKind, size32_t size)
+{
+    if (!isActive())
+        return;
+
+    size32_t requiredSize = sizeMessageHeaderFooter + getSizeOfAttrs(fileid, offset, nodeKind, size);
+    offset_type writeOffset = reserveEvent(requiredSize);
+    offset_type pos = writeOffset;
+    writeEventHeader(EventIndexEviction, pos);
+    write(pos, EvAttrFileId, fileid);
+    write(pos, EvAttrFileOffset, offset);
+    write(pos, EvAttrNodeKind, nodeKind);
+    write(pos, EvAttrExpandedSize, size);
+    writeEventFooter(pos, requiredSize, writeOffset);
+}
+
+void EventRecorder::recordDaliConnect(const char * path, __uint64 id)
+{
+    if (!isActive())
+        return;
+
+    size32_t requiredSize = sizeMessageHeaderFooter + getSizeOfAttrs(path, id);
+    offset_type writeOffset = reserveEvent(requiredSize);
+    offset_type pos = writeOffset;
+    writeEventHeader(EventDaliConnect, pos);
+    write(pos, EvAttrPath, path);
+    write(pos, EvAttrConnectId, id);
+    writeEventFooter(pos, requiredSize, writeOffset);
+}
+
+void EventRecorder::recordDaliDisconnect(__uint64 id)
+{
+    if (!isActive())
+        return;
+
+    size32_t requiredSize = sizeMessageHeaderFooter + getSizeOfAttrs(id);
+    offset_type writeOffset = reserveEvent(requiredSize);
+    offset_type pos = writeOffset;
+    writeEventHeader(EventDaliConnect, pos);
+    write(pos, EvAttrConnectId, id);
+    writeEventFooter(pos, requiredSize, writeOffset);
+}
+
+void EventRecorder::writeEventHeader(EventType type, offset_type & offset)
+{
+    __uint64 ts = cycle_to_nanosec(get_cycles_now() - startCycles); // nanoseconds relative to the start of the recording
+
+    write(offset, type);
+    write(offset, ts);
+
+    if (options & ERFtraceid)
+    {
+        const char * traceid = queryThreadedActiveSpan()->queryTraceId();
+        assertex(strlen(traceid) == 32);
+        writeData(offset, 32, traceid);
+    }
+    if (options & ERFthreadid)
+    {
+        __uint64 threadId = (__uint64)GetCurrentThreadId();
+        write(offset, threadId);
+    }
+}
+
+void EventRecorder::writeEventFooter(offset_type & offset, size32_t requiredSize, offset_t writeOffset)
+{
+    EventAttr endOfEvent = EvAttrNone;
+
+    write(offset, endOfEvent);
+    commitEvent(writeOffset, requiredSize);
+
+    //Sanity check after the event is committed to avoid potential deadlock
+    assertex(offset == writeOffset + requiredSize);
```

Comment:
perhaps should be a dbgassertex, i.e. no check in release?

Reply:
The cost is trivial, so I will keep as-is for the moment.